### PR TITLE
s390x: build s390x binary and image

### PR DESCRIPTION
**Pull request by huoqifeng**
## Description

Fixes: https://github.com/tigera/operator/issues/2603

## For PR author

- [ ] Tests for change.
- [ ] If changing pkg/apis/, run `make gen-files`
- [ ] If changing versions, run `make gen-versions`

## For PR reviewers

A note for code reviewers - all pull requests must have the following:

- [ ] Milestone set according to targeted release.
- [ ] Appropriate labels:
  - `kind/bug` if this is a bugfix.
  - `kind/enhancement` if this is a a new feature.
  - `enterprise` if this PR applies to Calico Enterprise only.

### CODE REVIEW

**Comment by huoqifeng** [comment]
@caseydavenport , I'm trying to build the s390x bin and image for the operator, would you please help have a look the PR and trigger CI for it?

**Comment by huoqifeng** [comment]
I can build the s390x binary and image on a amd64 box via command like below with the code change
```
make register build image ARCH=s390x
...
 ---> 9f00d4a5e0a9
Step 20/20 : USER ${USER_UID}
 ---> [Warning] The requested image's platform (linux/s390x) does not match the detected host platform (linux/amd64) and no specific platform was requested
 ---> Running in d31c7b25974c
Removing intermediate container d31c7b25974c
 ---> a070e950c9d9
Successfully built a070e950c9d9
Successfully tagged tigera/operator:latest-s390x
```

File: build/Dockerfile.s390x [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+RUN mkdir /licenses
+COPY LICENSE /licenses
+
+FROM scratch
+COPY --from=qemu /usr/bin/qemu-*-static /usr/bin/
+
+COPY --from=ubi /licenses /licenses
+COPY --from=ubi /etc/pki /etc/pki
+COPY --from=ubi /usr/share/pki /usr/share/pki
+COPY --from=ubi /lib /lib
```
Comment (matthewdupre) [inline build/Dockerfile.s390x:37]: The existing amd64 image copies a much smaller subset of libraries in - only 3 that are specifically needed.  This helps reduce the number of CVEs found in the image.  Do we need all of these libraries?

File: build/Dockerfile.s390x [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+# Update base packages to get security updates.
+RUN microdnf update
+
+# Add in top-level license file
+RUN mkdir /licenses
+COPY LICENSE /licenses
+
+FROM scratch
+COPY --from=qemu /usr/bin/qemu-*-static /usr/bin/
```
Comment (matthewdupre) [inline build/Dockerfile.s390x:32]: I see that this line exists in the arm64 image as well, but I'm curious if we need this in the final image or if it's just compile time?

Comment (huoqifeng) [reply build/Dockerfile.s390x:32]: I think we need copy `qemu-*-static` to scratch to identify the image `arch`.

File: build/Dockerfile.s390x (hunk 1 above)
Comment (huoqifeng) [reply build/Dockerfile.s390x:37]: Right, looks we don't need the .so as CGO_ENABLED=0 set for s390x.

**Review by caseydavenport — APPROVED** [review]

**Comment by caseydavenport** [comment]
@huoqifeng looks like the build timed out after 45 minutes - is it meant to take that long?

**Comment by huoqifeng** [comment]
> @huoqifeng looks like the build timed out after 45 minutes - is it meant to take that long?

@caseydavenport crossplatform build took longer time in my machine indeed. It's about (more than) 1 hour for me.

**Comment by huoqifeng** [comment]
Just run command `make register build image ARCH=s390x` on my x86 ubuntu 20.04 again, it tool 30 minutes.

**Comment by huoqifeng** [comment]
@caseydavenport I don't have permission to open https://tigera.semaphoreci.com/workflows/c30b33cc-843b-4800-8fba-e19ad97499d0?pipeline_id=93f53129-7d55-4882-b976-b0b26aee16da, is there any other errors besides timeout?

**Comment by lwr20** [comment]
I've copied the run output here: 
https://gist.github.com/lwr20/92cebc517ba755ea2318252d01119d00

**Comment by huoqifeng** [comment]
I saw the log from L2614 is like below:
```
command-line-arguments
cache store bin-s390x-${SEMAPHORE_GIT_SHA} build/_output/bin
Compressing build/_output/bin...
Compression complete. Duration: 10.858909971s. Size: 32.3M bytes.
Uploading 'build/_output/bin' with cache key 'bin-s390x-5447fac230298bbf2a4cd9f2cdfc3d6818e43f4f'...
Upload complete. Duration: 1.082847646s.
cache store go-pkg-cache-s390x-${SEMAPHORE_GIT_SHA} .go-pkg-cache
Compressing .go-pkg-cache...
Compression complete. Duration: 3m36.672387119s. Size: 264.3M bytes.
Uploading '.go-pkg-cache' with cache key 'go-pkg-cache-s390x-5447fac230298bbf2a4cd9f2cdfc3d6818e43f4f'...
```
From my local the log looks like this:
```
github.com/tigera/operator/pkg/awssgsetup
command-line-arguments
docker build --pull -t tigera/operator:latest-s390x --platform=linux/s390x --build-arg GIT_VERSION=ba224b2528ab -f ./build/Dockerfile.s390x .
Sending build context to Docker daemon  61.44MB
Step 1/20 : ARG QEMU_IMAGE=calico/go-build:v0.84-s390x
Step 2/20 : FROM ${QEMU_IMAGE} as qemu
v0.84-s390x: Pulling from calico/go-build
Digest: sha256:e5f7b9307a2d80316fabde50be40b95dd570c99762cae4aded92e67d8f525b49
Status: Image is up to date for calico/go-build:v0.84-s390x
 ---> b11aeb78df72
Step 3/20 : FROM registry.access.redhat.com/ubi8/ubi-minimal:8.7 AS ubi
8.7: Pulling from ubi8/ubi-minimal
```
@caseydavenport @lwr20 any idea what's wrong?

**Comment by huoqifeng** [comment]
@caseydavenport I increased the timeout, may you please kick off the CI?

**Comment by huoqifeng** [comment]
> /sem-approve

Thank you @lwr20 kick off the PR's CI, it looks good, are you OK to merge it? CC @caseydavenport

**Pull request merged**